File: src/page_fetcher.py
"""
Page fetcher module for retrieving web content using Selenium.

Originally inspired by m92vyas/llm-reader.
"""
import asyncio
from selenium import webdriver
from selenium.webdriver.chrome.options import Options


async def get_page_source(url: str,
                         wait: float = 1.5,
                         headless: bool = True,
                         user_agent: str = "Mozilla/5.0 (Windows Phone 10.0; Android 4.2.1; Microsoft; Lumia 640 XL LTE) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Mobile Safari/537.36 Edge/12.10166"
                         ) -> str:
    """
    Get HTML source using Selenium WebDriver.

    Args:
        url (str): The URL from which HTML content is to be extracted
        wait (float): Time to implicitly wait for the website to load. Default is 1.5 sec.
        headless (bool): Use headless browser or not. Default True
        user_agent (str): User agent string for the browser

    Returns:
        str: HTML source code of the page

    Raises:
        Exception: If there's an error while fetching the page
    """
    driver = None
    try:
        # Check if using Google Colab
        using_colab = False
        try:
            import google.colab
            using_colab = True
        except ImportError:
            using_colab = False

        # Configure Chrome options
        options = webdriver.ChromeOptions()
        if headless:
            options.add_argument('--headless')
        options.add_argument(f'--user-agent={user_agent}')
        
        # Add Colab-specific options if needed
        if using_colab:
            options.add_argument('--no-sandbox')
            options.add_argument('--disable-dev-shm-usage')

        # Create driver and fetch page
        driver = webdriver.Chrome(options=options)
        driver.get(url)
        driver.implicitly_wait(wait)

        return driver.page_source

    except Exception as e:
        raise Exception(f"Error while fetching page: {e}. Ensure Chrome and ChromeDriver are installed.")
    
    finally:
        if driver:
            driver.quit()

File: src/lib/pageFetcher.js
/**
 * Page fetcher module for retrieving web content using Puppeteer.
 * 
 * JavaScript/Node.js equivalent of the Python page_fetcher.py module.
 * Originally inspired by m92vyas/llm-reader.
 */

const puppeteer = require('puppeteer');

/**
 * Get HTML source using Puppeteer.
 * 
 * @param {string} url - The URL from which HTML content is to be extracted
 * @param {Object} options - Configuration options
 * @param {number} [options.wait=1.5] - Time to wait for the website to load (seconds)
 * @param {boolean} [options.headless=true] - Use headless browser or not
 * @param {string} [options.userAgent] - User agent string for the browser
 * @param {boolean} [options.showBrowser=false] - Show browser window (opposite of headless)
 * @param {boolean} [options.disableWebSecurity=false] - Disable web security (CORS) - use with caution
 * @returns {Promise<string>} HTML source code of the page
 * @throws {Error} If there's an error while fetching the page
 */
async function getPageSource(url, options = {}) {
    const {
        wait = 1.5,
        headless = true,
        showBrowser = false,
        disableWebSecurity = false,
        userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
    } = options;

    // showBrowser overrides headless setting
    const isHeadless = showBrowser ? false : headless;

    let browser = null;
    try {
        // Launch browser with secure Chrome options
        const args = [
            '--disable-dev-shm-usage', // Required for Docker/CI environments
            '--disable-features=VizDisplayCompositor'
        ];
        
        // Only disable web security if explicitly requested (use with caution)
        if (disableWebSecurity) {
            args.push('--disable-web-security');
        }
        
        browser = await puppeteer.launch({
            headless: isHeadless ? 'new' : false,
            args
        });

        const page = await browser.newPage();
        
        // Set modern viewport for better compatibility
        await page.setViewport({ width: 1920, height: 1080 });
        
        // Set user agent
        await page.setUserAgent(userAgent);
        
        // Navigate to the URL and wait for DOM content to load
        await page.goto(url, { 
            waitUntil: 'domcontentloaded',
            timeout: 30000 // 30 second timeout
        });
        
        // Wait for the specified time (equivalent to Python's implicit wait)
        if (wait > 0) {
            await new Promise(resolve => setTimeout(resolve, wait * 1000));
        }
        
        // Wait for page to be fully loaded (handles dynamic content)
        try {
            await page.waitForFunction(
                () => document.readyState === 'complete',
                { timeout: Math.max(wait * 1000, 5000) } // Minimum 5 seconds for readyState check
            );
        } catch (timeoutError) {
            // Continue anyway if timeout - some content is better than none
            console.warn('Page load timeout, continuing with partial content');
        }
        
        // Get the full HTML source
        return await page.content();

    } catch (error) {
        throw new Error(`Error while fetching page: ${error.message}. Ensure Chrome/Chromium is installed.`);
    } finally {
        if (browser) {
            await browser.close();
        }
    }
}

module.exports = {
    getPageSource
};